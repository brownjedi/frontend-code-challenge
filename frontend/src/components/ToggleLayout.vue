<template>
  <div :class="styles.container">
    <div
      :class="[isList ? styles.list : styles.card, animated && styles.gelatine, styles.inner]"
      @click.stop="handleClick"
      @animationend="animated = false"
    ></div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, useCssModule } from 'vue'

export default defineComponent({
  props: {
    isList: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['click'],
  setup(_, { emit }) {
    const styles = useCssModule()
    const animated = ref(false)

    function handleClick() {
      animated.value = true
      emit('click')
    }

    return { styles, handleClick, animated }
  },
})
</script>

<style lang="scss" module>
.container {
  display: flex;
  align-items: flex-end;
  cursor: pointer;

  .inner {
    width: 1.5rem;
    height: 1.5rem;
    padding: 0.25rem;
  }

  & > * {
    flex: 1;
  }

  .list {
    background: url('/img/show-data--cards.svg') no-repeat;
  }

  .card {
    background: url('/img/thumbnail--2.svg') no-repeat;
  }

  .gelatine {
    animation: gelatine 0.5s;
  }

  @keyframes gelatine {
    from,
    to {
      transform: scale(1, 1);
    }
    25% {
      transform: scale(0.9, 1.1);
    }
    50% {
      transform: scale(1.1, 0.9);
    }
    75% {
      transform: scale(0.95, 1.05);
    }
  }
}
</style>
