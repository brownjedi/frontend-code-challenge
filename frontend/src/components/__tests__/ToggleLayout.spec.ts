import { shallowMount } from '@vue/test-utils'
import ToggleLayout from '@/components/ToggleLayout.vue'

describe('components/ToggleSkeleton.vue', () => {
  it('has the expected html structure', () => {
    const wrapper = shallowMount(ToggleLayout)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('emits when clicked', async () => {
    const wrapper = shallowMount(ToggleLayout, {
      props: {
        isList: true,
      },
    })

    expect(wrapper.emitted().click).toBeUndefined()

    await wrapper.find('.inner').trigger('click')

    expect(wrapper.emitted().click.length).toEqual(1)
  })
})
