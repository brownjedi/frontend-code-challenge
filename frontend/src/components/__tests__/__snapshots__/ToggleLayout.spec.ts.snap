// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ToggleSkeleton.vue has the expected html structure 1`] = `
<div
  class="container"
>
  <div
    class="card inner"
  />
</div>
`;
